readme.txt
=== Receiptful for WooCommerce ===
Contributors: receiptful
Tags: receiptful, receiptful woocommerce, receipt, receipts, invoice, email, order confirmation, order mail, ecommerce mail
Requires at least: 3.6.0
Tested up to: 4.1
Stable tag: 1.1.3
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

Receiptful replaces and supercharges the default WooCommerce receipts. Just activate, add your API key and be awesome.

== Description ==

= What you would experience =

**Increased customer lifetime value**
We're sending out some beautiful emails combined with a good upsell for your customer. Offering a deal to your existing customers will increase their average lifetime!

**Increasing revenue**
Because the customer loved **your** service, and gets a great deal for their next order, they will soon order again! Giving you extra revenue.

**Setup in less than 10 minutes**
As described below, setup is easy and quick! 10 minutes for increased revenue? *Who wouldn't want that?!*

**Why FREE?**
You can read about why we made Receiptful free here: [Taking a Moonshot](https://receiptful.com/blog/taking-a-moonshot/)

**Quick setup**
Installing Receiptful is simple.

1. Download & Activate this plugin
2. Sign up for a free Receiptful account
3. Paste your API key in your site
4. Add your personal branding to the email & choose your Upsell

Simple huh?!

**Need help?**
[Open a support ticket](https://wordpress.org/support/plugin/receiptful-for-woocommerce), We're here to make your life easier!

**Please Note:** You require a Receiptful account ([sign up here for a 100% **FREE** account](http://receiptful.com))

[youtube https://www.youtube.com/watch?v=ae3eG7CqUrA&modestbranding=1&controls=0&autohide=1&wmode=transparent&enablejsapi=1&html5=1&rel=0&vq=hd1080]

== Installation ==

1. Go to the **Plugins > Add New page** in your WordPress admin.
2. Search for "Receiptful" and install the "Receiptful for WooCommerce" plugin.
3. Click **Activate Plugin**.
4. Go to **WooCommerce > Settings > Receiptful** and enter your Receiptful API key. (If you haven't signed up for Receiptful yet, now's the time!)

That's it! You can now customize your WooCommerce receipts and add upsells via your Receiptful dashboard.

== Frequently Asked Questions ==

= Do I need to modify any code? =

Nope - we take care of everything for you. Just install the plugin, add your API key and you’ll be good to go!

= Does Receiptful work with my theme(s)? =

Yes, Receiptful works with any theme - whether free, commercial or custom. You do however need WooCommerce activated for Receiptful to work.

== Screenshots ==

1. Receipt comparison WooCommerce VS. Receiptful
2. WordPress / WooCommerce Plugin Settings. Just add API key.
3. Receiptful's dashboard and mission control center.
4. Customizing your Receiptful template.
5. Add & activating upsells to be included in your receipts.
6. Example of customer profile overview with rich meta data.


== Changelog ==

= 1.1.3 - 01/04/2015 =

* [Fix] - Typo in filter name 'receiptful_api_args_related_products'
* [Improvement] - Prevent shipping coupons from having discount amounts
* [Improvement] - Prevent getting related products in the initial product sync
* [Improvement] - Automatically picking up Tax/totals translation from WooCommerce
* [Improvement] - Prevent notice when API doesn't return the 'products' parameter

= 1.1.2 - 12/03/2015 =

* [Add] - Receipt sync for better recommendations
* [Add] - Order note support
* [Add] - Product note support
* [Improvement] - Changed 'Shipping' to the actual shipping title
* [Improvement] - Changed textdomain to 'receiptful' for consistancy
* [Prevent] - Prevent notice in upcoming Receiptful update

= 1.1.1 - 05/03/2015 =

* [Add] - Product sync for better recommendations
* [Fix] - load translation files
* [Improvement] - Subtotals refactor
* [Improvement] - CDN for JavaScript - Improving loading time
* [Improvement] - Small queue improvements (don't add 400 response to queue)
* [Improvement] - Subscriptions email notifications

= 1.1.0 - 28/01/2015 =

* [Add] - Unit tests!
* [Add] - WooCommerce 2.3 support
* [Add] - Filters & hooks for extending/modifying
* [Add] - Receipt comparison screenshot, you should see it ;-)
* [Improvement] - Payment method to the receipt
* [Improvement] - Date parameter to the API call to keep order date/time equal
* [Improvement] - Support for multiple download URLs
* [Improvement] - Split up compatibility files in separate file
* [Improvement] - Email class refactor

= 1.0.5 - 14/01/2015 =

* [Happy New Year!]
* [Improvement] - Refactored email WC overrides
* [Fix] - Warning when descriptions < 100 char

= 1.0.4 - 18/12/2014 =

* [Add] - helper function to not copy meta data for subscription renewals
* [Add] - Send random products as related when none are found by core function
* [Improvement] - Sending discounts as negative number to API
* [Improvement] - Refactored helper functions
* [Fix] - Error when updating WooCommerce version while Receiptful is active
* [Fix] - for WC Subscriptions emails

= 1.0.3 - 12/12/2014 =

* [Add] - Support for product meta
* [Add] - Support for downloadable products (download links in Receipt)
* [Improvement] - Change the coupon tracking to JS at checkout
* [Fix] - Bug that caused coupon product restrictions
* [Fix] - Javascript error on thank you page

= 1.0.2 - 03/12/2014 =

* [Add] - Receiptful is now FREE
* [Add] - Added reporting for email conversions
* [Improvement] - Refactor the API class
* [Improvement] - Refactor related products code
* [Improvement] - Add more code commenting
* [Improvement] - Remove custom API endpoint for coupons
* [Fix] - WC Pending email sending
* [Fix] - Email being sent for digital downloads

= 1.0.1 - 19/11/2014 =

* [Add] - Plugin screen shots + banner + icon
* [Add] - Coupon usage tracking
* [Add] - Option to restrict coupon usage to customer email
* [Add] - WooCommerce 2.1.X support
* [Improvement] - Change CRON from 60 to 15 minutes
* [Improvement] - WooCommerce activated check for both network activated and single site
* [Fix] - Notice when using Free shipping upsell
* [Fix] - Incorrect coupon expiry date

= 1.0.0 - 22/10/2014 =

* Initial Release


== Upgrade Notice ==



changelog.txt
*** Receiptful for WooCommerce ***

2015-04-01 Version 1.1.3
* [Fix] - Typo in filter name 'receiptful_api_args_related_products'
* [Improvement] - Prevent shipping coupons from having discount amounts
* [Improvement] - Prevent getting related products in the initial product sync
* [Improvement] - Automatically picking up Tax/totals translation from WooCommerce
* [Improvement] - Prevent notice when API doesn't return the 'products' parameter

2015-03-12 Version 1.1.2
* [Add] - Receipt sync for better recommendations
* [Add] - Order note support
* [Add] - Product note support
* [Improvement] - Changed 'Shipping' to the actual shipping title
* [Improvement] - Changed textdomain to 'receiptful' for consistancy

2015-03-05  Version 1.1.1
* [Add] - Product sync for better recommendations
* [Fix] - load translation files
* [Improvement] - Subtotals refactor
* [Improvement] - CDN for JavaScript - Improving loading time
* [Improvement] - Small queue improvements (don't add 400 response to queue)
* [Improvement] - Subscriptions email notifications

2015-01-28  Version 1.1.0
* [Add] - Unit tests!
* [Add] - WooCommerce 2.3 support
* [Add] - Filters & hooks for extending/modifying
* [Add] - Receipt comparison screenshot, you should see it ;-)
* [Improvement] - Payment method to the receipt
* [Improvement] - Date parameter to the API call to keep order date/time equal
* [Improvement] - Support for multiple download URLs
* [Improvement] - Split up compatibility files in separate file
* [Improvement] - Email class refactor

2015-01-14  Version 1.0.5
* [Happy New Year!]
* [Improvement] - Refactored email WC overrides
* [Fix] - Warning when descriptions > 100 char

2014-12-18  Version 1.0.4
* [Add] - helper function to not copy meta data for subscription renewals
* [Add] - Send random products as related when none are found by core function
* [Improvement] - Sending discounts as negative number to API
* [Improvement] - Refactored helper functions
* [Fix] - Error when updating WooCommerce version while Receiptful is active
* [Fix] - for WC Subscriptions emails

2014-12-12  Version 1.0.3
* [Add] - Support for product meta
* [Add] - Support for downloadable products (download links in Receipt)
* [Improvement] - Change the coupon tracking to JS at checkout
* [Fix] - Bug that caused coupon product restrictions
* [Fix] - Javascript error on thank you page

2014-12-04  Version 1.0.2
* [Add] - Receiptful is now FREE
* [Add] - Added reporting for email conversions
* [Improvement] - Refactor the API class
* [Improvement] - Refactor related products code
* [Improvement] - Add more code commenting
* [Improvement] - Remove custom API endpoint for coupons
* [Fix] - WC Pending email sending
* [Fix] - Email being sent for digital downloads

2014-11-19  Version 1.0.1
* [Add] - Plugin screens shots + banner + icon
* [Add] - Coupon usage tracking
* [Add] - Option to restrict coupon usage to customer email
* [Add] - WooCommerce 2.1.X support
* [Improvement] - Change CRON from 60 to 15 minutes
* [Improvement] - WooCommerce activated check for both network activated and single site
* [Fix] - Notice when using Free shipping upsell
* [Fix] - Incorrect coupon expiry date

2014-11-01  Version 1.0.0
* Initial Release


includes/class-receiptful-email.php
<?php
if ( ! defined( 'ABSPATH' ) ) exit; // Exit if accessed directly

/**
 * Class Receiptful_Email.
 *
 * Admin class.
 *
 * @class		Receiptful_Email
 * @version		1.0.0
 * @author		Receiptful
 */
class Receiptful_Email {


	/**
	 * Constructor.
	 *
	 * @since 1.0.0
	 */
	public function __construct() {

		$this->hooks();

	}


	/**
	 * Class hooks.
	 *
	 * @since 1.0.0
	 */
	public function hooks() {

		// Remove standard emails
		add_filter( 'woocommerce_email_classes', array( $this, 'update_woocommerce_email' ), 90 );

		// Add hook to send new email
		//add_action( 'woocommerce_order_status_completed', array( $this, 'send_transactional_email' ) );
		add_action( 'woocommerce_order_status_pending_to_processing', array( $this, 'send_transactional_email' ) );
		add_action( 'woocommerce_order_status_on-hold_to_processing', array( $this, 'send_transactional_email' ) );
		add_action( 'woocommerce_order_status_pending_to_completed', array( $this, 'send_transactional_email' ) );

		// Save card last 4, card type and customer IP for sending with receipt
		add_action( 'woocommerce_checkout_order_processed', array( $this, 'save_card_data' ), 90, 2 );

		// Add coupon if the Receiptful API returns an upsell
		add_action( 'receiptful_add_upsell', array( $this, 'create_coupon' ), 10, 2 );

		// Add 'View Receipt' button to the My Account page
		add_filter( 'woocommerce_my_account_my_orders_actions', array( $this, 'view_receipt_button' ), 9, 2 );

		// Add option to Order Actions meta box on the Edit Order admin page
		add_action( 'woocommerce_order_actions', array( $this, 'receiptful_order_actions' ) );

		// Order Action callback
		add_action( 'woocommerce_order_action_receiptful_send_receipt', array( $this, 'send_transactional_email' ), 60);

	}


	/**
	 * WC Emails.
	 *
	 * Remove the WooCommerce Completed Order and New Order emails and add
	 * Receiptful email in their place.
	 *
	 * @since 1.0.0
	 *
	 * @param	array $emails	List of existing/registered WC emails.
	 * @return	array			List of modified WC emails.
	 */
	public function update_woocommerce_email( $emails ) {

		// Remove triggers
		remove_action( 'woocommerce_order_status_completed_notification', array( $emails['WC_Email_Customer_Completed_Order'], 'trigger' ) );
		remove_action( 'woocommerce_order_status_pending_to_processing_notification', array( $emails['WC_Email_Customer_Processing_Order'], 'trigger' ) );
		remove_action( 'woocommerce_order_status_pending_to_on-hold_notification', array( $emails['WC_Email_Customer_Processing_Order'], 'trigger' ) );

		// Remove WC_Email_Customer_Processing_Order
		unset( $emails['WC_Email_Customer_Processing_Order'] );

		// Remove WC_Email_Customer_Completed_Order
		unset( $emails['WC_Email_Customer_Completed_Order'] );

		// Add the Receiptful Completed Order email
		$emails['WC_Email_Customer_Completed_Order'] = include plugin_dir_path( __FILE__ ) . 'emails/class-receiptful-email-customer-new-order.php';

		return $emails;

	}


	/**
	 * Send mail.
	 *
	 * Init the mailer and call our notification for completed order.
	 *
	 * @since 1.0.0
	 */
	public function send_transactional_email() {

		WC()->mailer();
		$args = func_get_args();
		do_action_ref_array( 'receiptful_order_status_processing_notification', $args );

	}


	/**
	 * Save CC.
	 *
	 * Save the last 4 digits of the used credit card.
	 *
	 * @since 1.0.0
	 *
	 * @param int	$order_id	Order ID to save the card data to.
	 * @param array	$posted		List of posted data.
	 */
	public function save_card_data( $order_id, $posted ) {

		// TODO: try a different route
		// most gateways don't collect card type, but the number
		// save IP address

		return;

	}


	/**
	 * Create coupon.
	 *
	 * Create a coupon when upsell data returned from Receiptful API.
	 *
	 * @since 1.0.0
	 *
	 * @param	array	$data		List of data returned by the Receiptful API.
	 * @param	int		$order_id	ID of the order being processed.
	 * @return  int     $id         ID of new coupon
	 */
	public function create_coupon( $data, $order_id ) {

		$order				= wc_get_order( $order_id );
		$coupon_code		= apply_filters( 'woocommerce_coupon_code', wc_clean( $data['couponCode'] ) );
		$shipping_coupon	= 'no';
		$discount_type		= 'fixed_cart';

		// Check for duplicate coupon codes
		$coupon_found = wc_get_coupon_by_code( $coupon_code );

		// Return the coupon ID when it already exists
		if ( $coupon_found ) {
			return $coupon_found;
		}

		$expiry_date = date_i18n( 'Y-m-d', strtotime( '+' . wc_clean( $data['expiryPeriod'] ) . ' day' ) );

		if ( 'discountcoupon' == $data['upsellType'] ) {

			switch ( wc_clean( $data['couponType'] ) ) {
				case 1:
					$discount_type = 'fixed_cart';
				break;
				case 2:
					$discount_type = 'percent';
				break;
				default:
					$discount_type = 'fixed_cart';
				break;
			}

		} elseif ( 'shippingcoupon' == $data['upsellType'] ) {

			$shipping_coupon 	= 'yes';
			$data['amount']		= '';

		}

		$coupon_data = apply_filters( 'receiptful_coupon_data', array(
			'type'							=> $discount_type,
			'amount'						=> isset( $data['amount'] ) ? wc_clean( $data['amount'] ) : '',
			'individual_use'				=> 'yes',
			'product_ids'					=> '',
			'exclude_product_ids'			=> array(),
			'usage_limit'					=> '1',
			'usage_limit_per_user'			=> '1',
			'limit_usage_to_x_items'		=> '',
			'usage_count'					=> '',
			'expiry_date'					=> $expiry_date,
			'apply_before_tax'				=> 'yes',
			'free_shipping'					=> $shipping_coupon,
			'product_categories'			=> array(),
			'exclude_product_categories'	=> array(),
			'exclude_sale_items'			=> 'no',
			'minimum_amount'				=> '',
			'maximum_amount'				=> '',
			'customer_email'				=> ! empty( $data['emailLimit'] ) ? array( $order->billing_email ) : array(),
		), $order_id, $data );

		$new_coupon = array(
			'post_title'	=> $coupon_code,
			'post_content'	=> '',
			'post_status'	=> 'publish',
			'post_author'	=> get_current_user_id(),
			'post_type'		=> 'shop_coupon',
			'post_excerpt'	=> isset( $data['title'] ) ? wc_clean( $data['title'] ) : '',
		);
		$id = wp_insert_post( $new_coupon );

		// set coupon meta
		update_post_meta( $id, 'discount_type', $coupon_data['type'] );
		update_post_meta( $id, 'coupon_amount', wc_format_decimal( $coupon_data['amount'] ) );
		update_post_meta( $id, 'individual_use', $coupon_data['individual_use'] );
		update_post_meta( $id, 'product_ids', '' );
		update_post_meta( $id, 'exclude_product_ids', implode( ',', array_filter( array_map( 'intval', $coupon_data['exclude_product_ids'] ) ) ) );
		update_post_meta( $id, 'usage_limit', absint( $coupon_data['usage_limit'] ) );
		update_post_meta( $id, 'usage_limit_per_user', absint( $coupon_data['usage_limit_per_user'] ) );
		update_post_meta( $id, 'limit_usage_to_x_items', absint( $coupon_data['limit_usage_to_x_items'] ) );
		update_post_meta( $id, 'usage_count', absint( $coupon_data['usage_count'] ) );
		update_post_meta( $id, 'expiry_date', wc_clean( $coupon_data['expiry_date'] ) );
		update_post_meta( $id, 'apply_before_tax', wc_clean( $coupon_data['apply_before_tax'] ) );
		update_post_meta( $id, 'free_shipping', wc_clean( $coupon_data['free_shipping'] ) );
		update_post_meta( $id, 'product_categories', array_filter( array_map( 'intval', $coupon_data['product_categories'] ) ) );
		update_post_meta( $id, 'exclude_product_categories', array_filter( array_map( 'intval', $coupon_data['exclude_product_categories'] ) ) );
		update_post_meta( $id, 'exclude_sale_items', wc_clean( $coupon_data['exclude_sale_items'] ) );
		update_post_meta( $id, 'minimum_amount', wc_format_decimal( $coupon_data['minimum_amount'] ) );
		update_post_meta( $id, 'maximum_amount', wc_format_decimal( $coupon_data['maximum_amount'] ) );
		update_post_meta( $id, 'customer_email', array_filter( array_map( 'sanitize_email', $coupon_data['customer_email'] ) ) );
		update_post_meta( $id, 'receiptful_coupon', 'yes' );
		update_post_meta( $id, 'receiptful_coupon_order', $order_id );

		return $id;

	}


	/**
	 * Online receipt.
	 *
	 * Prints the View Receipt button to view the receipt online.
	 *
	 * @since 1.0.0
	 *
	 * @param	array		$actions	List of existing actions (buttons).
	 * @param	WC_Order	$order		Order object of the current order line.
	 * @return	array					List of modified actions (buttons).
	 */
	public function view_receipt_button( $actions, $order ) {

		$receipt_id				= get_post_meta( $order->id, '_receiptful_receipt_id', true );
		$receiptful_web_link	= get_post_meta( $order->id, '_receiptful_web_link', true );

		if ( $receipt_id && $receiptful_web_link ){
			// Id exists so remove old View button and add Receiptful button
			unset( $actions['view'] );

			$actions['receipt'] = array(
				'url'	=> $receiptful_web_link['webview'],
				'name'	=> __( 'View Receipt', 'receiptful' )
			);
		}

		return $actions;

	}


	/**
	 * Resend queue.
	 *
	 * Resend receipts in queue. Called from Cron.
	 *
	 * @since 1.0.0
	 */
	public static function resend_queue(){

		// Check queue
		$resend_queue = get_option( '_receiptful_resend_queue' );

		if ( is_array( $resend_queue ) && ( count( $resend_queue ) > 0 ) ) {

			WC()->mailer();
			foreach ( $resend_queue as $key => $order_id ) {

				$receiptful_email 	= new Receiptful_Email_Customer_New_Order();
				$response 			= $receiptful_email->trigger( $order_id );

				if ( ! is_wp_error( $response ) && in_array( $response['response']['code'], array( '200', '201', '400' ) ) ) {
					unset( $resend_queue[ $key ] );
				}

			}

			update_option( '_receiptful_resend_queue', $resend_queue );

		}

	}


	/**
	 * Order actions.
	 *
	 * Display the Receiptful action in the Order Actions meta box drop down.
	 *
	 * @since 1.0.0
	 *
	 * @param	array $actions	List of existing order actions.
	 * @return	array			List of modified order actions.
	 */
	public function receiptful_order_actions( $actions ) {

		if ( is_array( $actions ) ) {
			$actions['receiptful_send_receipt'] = __( 'Send receipt', 'receiptful' );
		}

		return $actions;

	}


}
